### ISSUE TITLE
VF configured/attached with unsupported operations in net-attach-def doesn't gets de-configured on pod deletion

### ISSUE THREAD

**Issue by mukesh-dua** [k8snetworkplumbingwg/sriov-cni#119]
**Problem Statement:** If an unsupported operation is added to network attachment definition, it allows for configuration of Virtual Function during pod creation and the configuration remains even on pod deletion. 

The pod never gets created with such a network attachment definition. The end result is someone has to delete the pod.

This is a problem in multi-tenant environment where tenants may be given rights to configure network attachment definitions and they may trying configuring some settings which can leave VF in an undesired state.

**Expectations:**
On pod deletion, the configurations that were set using net-attach-defintions on Virtual functions are reverted back.

**Steps to reproduce:**
```
$ #Get the pool size for resource name - intel_sriov_dpdk
$ kubectl describe nodes compute-1 | grep -A15 Capacity
Capacity:
 cpu:                              16
 ephemeral-storage:                238721632Ki
 hugepages-1Gi:                    210Gi
 intel.com/intel_sriov_dpdk:       16
 intel.com/intel_sriov_netdevice:  8
 memory:                           264162932Ki
 pods:                             110
Allocatable:
 cpu:                              16
 ephemeral-storage:                220005855687
 hugepages-1Gi:                    210Gi
 intel.com/intel_sriov_dpdk:       16
 intel.com/intel_sriov_netdevice:  8
 memory:                           43859572Ki
 pods:                             110
$
$ #Current resource allocations
$ kubectl describe nodes compute-1 | grep -A9 Allocated
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                         Requests   Limits
  --------                         --------   ------
  cpu                              350m (2%)  100m (0%)
  memory                           50Mi (0%)  50Mi (0%)
  ephemeral-storage                0 (0%)     0 (0%)
  intel.com/intel_sriov_dpdk       1          1
  intel.com/intel_sriov_netdevice  1          1
Events:                            <none>
$
$ #current VF assignments - before creation of faulty pod
$ ip link show enp68s0f0
4: enp68s0f0: <BROADCAST,MULTICAST> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether a0:36:9f:26:f9:30 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 8 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 9 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 10 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 11 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 12 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 13 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 14 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 15 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 16 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 17 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 18 MAC 00:00:00:00:00:00, vlan 1000, spoof checking off, link-state auto, trust on, query_rss off
    vf 19 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 20 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 21 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 22 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 23 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 24 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 25 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 26 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 27 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 28 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 29 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 30 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 31 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
$
$ #Definition of network attachment definition - with link state (probably not a supported function on ixgbevf)
$ cat intel-sriov-dpdk-nad-faulty.yaml 
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: intel-sriov-dpdk-nad-faulty
  annotations:
    k8s.v1.cni.cncf.io/resourceName: intel.com/intel_sriov_dpdk
spec:
  config: '{
  "type": "sriov",
  "cniVersion": "0.3.1",
  "name": "sriov-network",
  "vlan": 1000,
  "spoofchk": "off",
  "trust": "on",
  "link_state": "enable"
}'
$
$ #Definition of pod using the network attachment definition
$ cat sriov-testpod-dpdk-faulty.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: sriov-testpod-dpdk-faulty
  annotations:
    k8s.v1.cni.cncf.io/networks: intel-sriov-dpdk-nad-faulty 
spec:
  containers:
  - name: appcntr1 
    image: centos/tools 
    imagePullPolicy: IfNotPresent
    command: [ "/bin/bash", "-c", "--" ]
    args: [ "while true; do sleep 300000; done;" ]
    resources:
      requests:
        intel.com/intel_sriov_dpdk: '1'
      limits:
        intel.com/intel_sriov_dpdk: '1'
$
$ #Create network attachment definition
$ kubectl create -f intel-sriov-dpdk-nad-faulty.yaml 
networkattachmentdefinition.k8s.cni.cncf.io/intel-sriov-dpdk-nad-faulty created
$
$ #Get all network attachment defintion. The newly created with name "intel-sriov-dpdk-nad-faulty"
$ kubectl get net-attach-def
NAME                          AGE
intel-sriov-dpdk-nad-faulty   7s
sriov-dev-net1                23h
sriov-dpdk-net1               23h
$
$ #Create a pod with intel-sriov-dpdk-nad-faulty network attachment definition
$ kubectl create -f sriov-testpod-dpdk-faulty.yaml 
pod/sriov-testpod-dpdk-faulty created
$
$ #Get the status of all the pods
$ kubectl get po -A | grep sriov
default        sriov-testpod-dpdk                                      1/1     Running             0          45m
default        sriov-testpod-dpdk-faulty                               0/1     ContainerCreating   0          10s
default        sriov-testpod1                                          1/1     Running             0          23h
kube-system    kube-sriov-cni-ds-amd64-8wmzf                           1/1     Running             0          24h
kube-system    kube-sriov-cni-ds-amd64-lc4hp                           1/1     Running             0          24h
kube-system    kube-sriov-device-plugin-amd64-kqhgz                    1/1     Running             0          23h
kube-system    kube-sriov-device-plugin-amd64-prpfj                    1/1     Running             0          8h
$
$ #describe the pod status 
$ kubectl describe po sriov-testpod-dpdk-faulty 
Name:         sriov-testpod-dpdk-faulty
Namespace:    default
Priority:     0
Node:         compute-1/70.151.43.232
Start Time:   Fri, 13 Mar 2020 02:21:01 +0000
Labels:       <none>
Annotations:  cni.projectcalico.org/podIP: 172.16.100.107/32
              k8s.v1.cni.cncf.io/networks: intel-sriov-dpdk-nad-faulty
              k8s.v1.cni.cncf.io/networks-status: 
Status:       Pending
IP:           
IPs:          <none>
Containers:
  appcntr1:
    Container ID:  
    Image:         centos/tools
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/bash
      -c
      --
    Args:
      while true; do sleep 300000; done;
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Limits:
      intel.com/intel_sriov_dpdk:  1
    Requests:
      intel.com/intel_sriov_dpdk:  1
    Environment:                   <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-pkm2n (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  default-token-pkm2n:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-pkm2n
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                  Age               From                  Message
  ----     ------                  ----              ----                  -------
  Normal   Scheduled               62s               default-scheduler     Successfully assigned default/sriov-testpod-dpdk-faulty to compute-1
  Warning  FailedCreatePodSandBox  54s               kubelet, compute-1  Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "412de788aeb04c5cd151c7815dfd16ccd99f65df35a738f65ad46e74ed23bb86" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to set up pod "sriov-testpod-dpdk-faulty_default" network: Multus: [default/sriov-testpod-dpdk-faulty]: error adding container to network "sriov-network": delegateAdd: error invoking DelegateAdd - "sriov": error in getting result from AddNetwork: SRIOV-CNI failed to configure VF "failed to set vf 28 link state to 1: operation not supported", failed to clean up sandbox container "412de788aeb04c5cd151c7815dfd16ccd99f65df35a738f65ad46e74ed23bb86" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to teardown pod "sriov-testpod-dpdk-faulty_default" network: delegateDel: error invoking DelegateDel - "sriov": error in getting result from DelNetwork: error reading cached NetConf in /var/lib/cni/sriov with name 412de788aeb04c5cd151c7815dfd16ccd99f65df35a738f65ad46e74ed23bb86-net1]
  Warning  FailedCreatePodSandBox  46s               kubelet, compute-1  Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "d8263d4b1f71eaefc02e5db7b2486dea86567564d2e156b20fe8ea65273498e8" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to set up pod "sriov-testpod-dpdk-faulty_default" network: Multus: [default/sriov-testpod-dpdk-faulty]: error adding container to network "sriov-network": delegateAdd: error invoking DelegateAdd - "sriov": error in getting result from AddNetwork: SRIOV-CNI failed to configure VF "failed to set vf 28 link state to 1: operation not supported", failed to clean up sandbox container "d8263d4b1f71eaefc02e5db7b2486dea86567564d2e156b20fe8ea65273498e8" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to teardown pod "sriov-testpod-dpdk-faulty_default" network: delegateDel: error invoking DelegateDel - "sriov": error in getting result from DelNetwork: error reading cached NetConf in /var/lib/cni/sriov with name d8263d4b1f71eaefc02e5db7b2486dea86567564d2e156b20fe8ea65273498e8-net1]
  Warning  FailedCreatePodSandBox  38s               kubelet, compute-1  Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "3adb32a0702d6476dca9d106c96610c756517e8abf6266cd2408231a3479df6e" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to set up pod "sriov-testpod-dpdk-faulty_default" network: Multus: [default/sriov-testpod-dpdk-faulty]: error adding container to network "sriov-network": delegateAdd: error invoking DelegateAdd - "sriov": error in getting result from AddNetwork: SRIOV-CNI failed to configure VF "failed to set vf 28 link state to 1: operation not supported", failed to clean up sandbox container "3adb32a0702d6476dca9d106c96610c756517e8abf6266cd2408231a3479df6e" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to teardown pod "sriov-testpod-dpdk-faulty_default" network: delegateDel: error invoking DelegateDel - "sriov": error in getting result from DelNetwork: error reading cached NetConf in /var/lib/cni/sriov with name 3adb32a0702d6476dca9d106c96610c756517e8abf6266cd2408231a3479df6e-net1]
  Warning  FailedCreatePodSandBox  30s               kubelet, compute-1  Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "a239769ec3b60cf7db81d590a996165b95ab23e75cbd17a578bd067508cef678" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to set up pod "sriov-testpod-dpdk-faulty_default" network: Multus: [default/sriov-testpod-dpdk-faulty]: error adding container to network "sriov-network": delegateAdd: error invoking DelegateAdd - "sriov": error in getting result from AddNetwork: SRIOV-CNI failed to configure VF "failed to set vf 28 link state to 1: operation not supported", failed to clean up sandbox container "a239769ec3b60cf7db81d590a996165b95ab23e75cbd17a578bd067508cef678" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to teardown pod "sriov-testpod-dpdk-faulty_default" network: delegateDel: error invoking DelegateDel - "sriov": error in getting result from DelNetwork: error reading cached NetConf in /var/lib/cni/sriov with name a239769ec3b60cf7db81d590a996165b95ab23e75cbd17a578bd067508cef678-net1]
  Warning  FailedCreatePodSandBox  22s               kubelet, compute-1  Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "46abcd7d2d18a8d1b90ee2764ab92626df67a15bed0a3171f479f6641b23ba56" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to set up pod "sriov-testpod-dpdk-faulty_default" network: Multus: [default/sriov-testpod-dpdk-faulty]: error adding container to network "sriov-network": delegateAdd: error invoking DelegateAdd - "sriov": error in getting result from AddNetwork: SRIOV-CNI failed to configure VF "failed to set vf 28 link state to 1: operation not supported", failed to clean up sandbox container "46abcd7d2d18a8d1b90ee2764ab92626df67a15bed0a3171f479f6641b23ba56" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to teardown pod "sriov-testpod-dpdk-faulty_default" network: delegateDel: error invoking DelegateDel - "sriov": error in getting result from DelNetwork: error reading cached NetConf in /var/lib/cni/sriov with name 46abcd7d2d18a8d1b90ee2764ab92626df67a15bed0a3171f479f6641b23ba56-net1]
  Warning  FailedCreatePodSandBox  13s               kubelet, compute-1  Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "5835a83db2c1b6f24a97e1d2ea901b4163459d54d99a37b5d153fe634d91243d" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to set up pod "sriov-testpod-dpdk-faulty_default" network: Multus: [default/sriov-testpod-dpdk-faulty]: error adding container to network "sriov-network": delegateAdd: error invoking DelegateAdd - "sriov": error in getting result from AddNetwork: SRIOV-CNI failed to configure VF "failed to set vf 28 link state to 1: operation not supported", failed to clean up sandbox container "5835a83db2c1b6f24a97e1d2ea901b4163459d54d99a37b5d153fe634d91243d" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to teardown pod "sriov-testpod-dpdk-faulty_default" network: delegateDel: error invoking DelegateDel - "sriov": error in getting result from DelNetwork: error reading cached NetConf in /var/lib/cni/sriov with name 5835a83db2c1b6f24a97e1d2ea901b4163459d54d99a37b5d153fe634d91243d-net1]
  Warning  FailedCreatePodSandBox  5s                kubelet, compute-1  Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "0e1d5bd3d2cdd128e3cfe33959881a7065605de37cc3319e3f489641ecefa720" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to set up pod "sriov-testpod-dpdk-faulty_default" network: Multus: [default/sriov-testpod-dpdk-faulty]: error adding container to network "sriov-network": delegateAdd: error invoking DelegateAdd - "sriov": error in getting result from AddNetwork: SRIOV-CNI failed to configure VF "failed to set vf 28 link state to 1: operation not supported", failed to clean up sandbox container "0e1d5bd3d2cdd128e3cfe33959881a7065605de37cc3319e3f489641ecefa720" network for pod "sriov-testpod-dpdk-faulty": networkPlugin cni failed to teardown pod "sriov-testpod-dpdk-faulty_default" network: delegateDel: error invoking DelegateDel - "sriov": error in getting result from DelNetwork: error reading cached NetConf in /var/lib/cni/sriov with name 0e1d5bd3d2cdd128e3cfe33959881a7065605de37cc3319e3f489641ecefa720-net1]
  Normal   SandboxChanged          4s (x7 over 54s)  kubelet, compute-1  Pod sandbox changed, it will be killed and re-created.
$
$ #pod stucks in ContainerCreating state with error reason "ailed to set vf 28 link state to 1: operation not supported"
$
$ #Check the state of Virtual Functions - vlan 28 is now configured with vlan 1000, trust mode on and spoof check off
$ ip link show enp68s0f0
4: enp68s0f0: <BROADCAST,MULTICAST> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether a0:36:9f:26:f9:30 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 8 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 9 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 10 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 11 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 12 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 13 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 14 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 15 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 16 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 17 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 18 MAC 00:00:00:00:00:00, vlan 1000, spoof checking off, link-state auto, trust on, query_rss off
    vf 19 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 20 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 21 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 22 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 23 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 24 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 25 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 26 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 27 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 28 MAC 00:00:00:00:00:00, vlan 1000, spoof checking off, link-state auto, trust on, query_rss off
    vf 29 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 30 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 31 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
$
$ #Get the current allocated pool 
$ kubectl describe nodes compute-1 | grep -A9 Allocated
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                         Requests   Limits
  --------                         --------   ------
  cpu                              350m (2%)  100m (0%)
  memory                           50Mi (0%)  50Mi (0%)
  ephemeral-storage                0 (0%)     0 (0%)
  intel.com/intel_sriov_dpdk       2          2
  intel.com/intel_sriov_netdevice  1          1
Events:                            <none>
```

Now, even after pod deletion the settings for VF are still intact.
```
After pod deletion:
==================
$ Deleting the faulty pod
$ kubectl delete po sriov-testpod-dpdk-faulty
pod "sriov-testpod-dpdk-faulty" deleted
$
$ #pod has been deleted now
$ kubectl get po -A | grep sriov
default        sriov-testpod-dpdk                                      1/1     Running            0          67m
default        sriov-testpod1                                          1/1     Running            0          24h
kube-system    kube-sriov-cni-ds-amd64-8wmzf                           1/1     Running            0          24h
kube-system    kube-sriov-cni-ds-amd64-lc4hp                           1/1     Running            0          24h
kube-system    kube-sriov-device-plugin-amd64-kqhgz                    1/1     Running            0          24h
kube-system    kube-sriov-device-plugin-amd64-prpfj                    1/1     Running            0          9h
$
$ #Check the status of VF. It still has vlan 1000 attached to it
$ ip link show enp68s0f0
4: enp68s0f0: <BROADCAST,MULTICAST> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether a0:36:9f:26:f9:30 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 1 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 5 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 6 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 7 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 8 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 9 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 10 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 11 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 12 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 13 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 14 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 15 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 16 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 17 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 18 MAC 00:00:00:00:00:00, vlan 1000, spoof checking off, link-state auto, trust on, query_rss off
    vf 19 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 20 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 21 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 22 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 23 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 24 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 25 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 26 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 27 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 28 MAC 00:00:00:00:00:00, vlan 1000, spoof checking off, link-state auto, trust on, query_rss off
    vf 29 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 30 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
    vf 31 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust off, query_rss off
$
$ #Allocated pool gets populated properly - incrementing dpdk pool on deletion of pod
$ kubectl describe nodes compute-1 | grep -A9 Allocated
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                         Requests   Limits
  --------                         --------   ------
  cpu                              350m (2%)  100m (0%)
  memory                           50Mi (0%)  50Mi (0%)
  ephemeral-storage                0 (0%)     0 (0%)
  intel.com/intel_sriov_dpdk       1          1
  intel.com/intel_sriov_netdevice  1          1
Events:                            <none>
```

Component | Version|
------------------------------|--------------------|
|SR-IOV CNI Plugin            |v2.2|
|Multus                       |v3.4|
|SR-IOV Network Device Plugin |v3.1|
|Kubernetes                   |v1.16.2| 
|OS                           |Ubuntu 18.04.1 kernel version 4.15.0-88-generic|

### Config Files
Config file locations may be config dependent - SR-IOV config
```
  config.json: |
    {
        "resourceList": [{
                "resourceName": "intel_sriov_dpdk",
                "selectors": {
                    "vendors": ["8086"],
                    "devices": ["10ed"],
                    "drivers": ["vfio-pci"],
                    "pfNames": ["enp67s0f1","enp68s0f0"]
                }
            }
        ]
    }
```

##### Multus config (Try '/etc/cni/multus/net.d')
```
{
  "name": "multus-cni-network",
  "type": "multus",
  "cniVersion": "0.3.1",
  "kubeconfig": "/etc/cni/net.d/multus.d/multus.kubeconfig",
  "confDir": "/etc/cni/net.d",
  "cniDir": "/var/lib/cni/multus",
  "binDir": "/opt/cni/bin",
  "logFile": "/var/log/multus.log",
  "logLevel": "panic",
  "capabilities": {
    "portMappings": true
  },
  "readinessindicatorfile": "",
  "namespaceIsolation": false,
  "clusterNetwork": "k8s-pod-network",
  "defaultNetwork": [],
  "systemNamespaces": ["kube-system"]
}
```

**Comment by zshi-redhat** [comment]
@mukesh-dua thanks for detailed report!

I think this is an issue in sriov-cni that it failed to retrieve local cache for CNI netConf which results in early exit in CNI cmdDel call, which in turn causes resetting of VF config not executed. The fix would be to cache netConf in CNI cmdAdd before applying VF configs so that when it fails, it can still recover from pre-cached netConf.

I believe [PR](https://github.com/intel/sriov-cni/pull/114) might be able to fix link state specific issue.
but we may need another change to lift caching of netConf to the beginning part of cmdAdd to take care of other VF configs such as vlan, qos etc.

**Comment by mukesh-dua** [comment]
> @mukesh-dua thanks for detailed report!
> 
> I think this is an issue in sriov-cni that it failed to retrieve local cache for CNI netConf which results in early exit in CNI cmdDel call, which in turn causes resetting of VF config not executed. The fix would be to cache netConf in CNI cmdAdd before applying VF configs so that when it fails, it can still recover from pre-cached netConf.
> 
> I believe [PR](https://github.com/intel/sriov-cni/pull/114) might be able to fix link state specific issue.
> but we may need another change to lift caching of netConf to the beginning part of cmdAdd to unlock the call of resetVFConfig.

Thanks Zshi. BTW, are there any specific timelines by which we expect this functionality to be part of sriov-cni release?

**Comment by zshi-redhat** [comment]
@mukesh-dua I don't have a timeline for it, but we're still reviewing that PR and it looks promising.

**Comment by zshi-redhat** [comment]
@mukesh-dua the aforementioned PR is merged in SR-IOV CNI and released in version v2.3. Does it work for you?

**Comment by adrianchiris** [comment]
Greetings!
This issue is related to : #115 as cleaning up a VF which was partly configured is one of the open items.

Closing this issue as it will be handled as part of #115
